### PR TITLE
[Sdk 358] SDK Limits, App quit null pointer exception, 

### PR DESCRIPTION
**Pull request by ZahidZafar**

### CODE REVIEW

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
code unrelated to SDK limits should be moved to a different PR and branch

File: Assets/Tests/PlayModeTests/ConfigurationTests.cs [hunk 1]
```diff
@@ -34,6 +34,12 @@ public void TestSDKInitParams()
                 EventQueueThreshold = 150,
                 TotalBreadcrumbsAllowed = 200,
 
+                MaxValueSize = 5,
```
Comment (ArtursKadikis) [inline Assets/Tests/PlayModeTests/ConfigurationTests.cs:37]: In these tests have a different value for each size. This way you can make sure that you are setting the correct variable

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Limits should be added to all user profile operations

File: Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs [hunk 2]
```diff
@@ -84,6 +71,19 @@ public async Task SetUserDetailsAsync(CountlyUserDetailsModel userDetailsModel)
...
+                userDetailsModel.Name = TrimValue(userDetailsModel.Name);
+                userDetailsModel.Phone = TrimValue(userDetailsModel.Phone);
+                userDetailsModel.Email = TrimValue(userDetailsModel.Email);
+                userDetailsModel.Gender = TrimValue(userDetailsModel.Gender);
+                userDetailsModel.Username = TrimValue(userDetailsModel.Username);
+                userDetailsModel.BirthYear = TrimValue(userDetailsModel.BirthYear);
+                userDetailsModel.Organization = TrimValue(userDetailsModel.Organization);
+
+                if (userDetailsModel.PictureUrl.Length > 4096) {
+                    userDetailsModel.PictureUrl = userDetailsModel.PictureUrl.Substring(0, 4096);
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs:83]: No warning printed

File: Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs [hunk 3]
```diff
@@ -23,6 +24,99 @@ protected AbstractBaseService(CountlyConfiguration configuration, CountlyLogHelp
...
+            }
+            
+
+            return values;
+        }
+
+        protected string TrimValue(string v)
+        {
+            if (v.Length > _configuration.MaxValueSize) {
+                Log.Verbose("[" + GetType().Name + "] TrimValue : Max allowed value length is " + _configuration.MaxValueSize);
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs:88]: should be warning

File: Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs [hunk 4]
```diff
@@ -23,6 +24,99 @@ protected AbstractBaseService(CountlyConfiguration configuration, CountlyLogHelp
...
+            }
+
+            return k;
+        }
+
+        protected string[] TrimValues(string[] values)
+        {
+            for (int i = 0; i < values.Length; ++i) {
+                if (values[i].Length > _configuration.MaxValueSize) {
+                    Log.Verbose("[" + GetType().Name + "] TrimKey : Max allowed key length is " + _configuration.MaxKeyLength);
```
[the remark this answers is not in the document]
Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs:76]: and should be warning

File: Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs [hunk 5]
```diff
@@ -84,6 +71,19 @@ public async Task SetUserDetailsAsync(CountlyUserDetailsModel userDetailsModel)
                     throw new Exception("Accepted picture formats are .png, .gif and .jpeg");
                 }
 
+                userDetailsModel.Name = TrimValue(userDetailsModel.Name);
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs:74]: Every one of these should print a unique warning message if they are trimmed. That message should say which field was over the limit

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,8 @@
+## 20.11.5
+* Added new configuration fields.
```
Comment (ArtursKadikis) [inline CHANGELOG.md:2]: "Added new configuration fields to manipulate internal SDK value and key limits"

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,8 @@
...
+* Added SDK internal limits.
```
Comment (ArtursKadikis) [inline CHANGELOG.md:3]: not needed

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,8 @@
...
+* Reworked request logging.
```
Comment (ArtursKadikis) [inline CHANGELOG.md:4]: not needed

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Warning messages in "UserDetailsCountlyService.cs" should be reworked as they have multiple issies:
* there should be a space before "isn't"
* those messages are just not helpful
* both warning should explicitly mention that the provided information would not be recorded because the provided values are not valid
* the warning for the key check should also just say that it's either null or empty not just that it "isn't valid" as that provides little information
* the warning for the value check should also mention for which key that value is provided


All logs, that are printed when a key/value is too long, should mention that the key/value will be truncated.

Run code formatting on the whole project.

Add a test that verifies that the maximum picture url size is enforced correctly.

User profile "multiply" call is not part of the limit tests

A warning message should be added in the changelog about the added internal limits

File: Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs [hunk 9]
```diff
@@ -23,6 +24,99 @@ protected AbstractBaseService(CountlyConfiguration configuration, CountlyLogHelp
...
+                    && (item.Value.GetType() == typeof(int)
+                    || item.Value.GetType() == typeof(bool)
+                    || item.Value.GetType() == typeof(float)
+                    || item.Value.GetType() == typeof(double)
+                    || item.Value.GetType() == typeof(string));
+
+
+                if (!isValidDataType) {
+                    toRemove.Add(item.Key);
+                    Log.Warning("[" + GetType().Name + "] RemoveSegmentInvalidetDataTypes: In segmentation Data type '" + (item.Value?.GetType()) + "'  of item '" + item.Key + "' isn't valid.");
```
[the remark this answers is not in the document]
Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs:51]: good catch

File: Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs [hunk 10]
```diff
@@ -23,6 +24,99 @@ protected AbstractBaseService(CountlyConfiguration configuration, CountlyLogHelp
...
+        {
+            if (v.Length > _configuration.MaxValueSize) {
+                Log.Verbose("[" + GetType().Name + "] TrimValue : Max allowed value length is " + _configuration.MaxValueSize);
+                v = v.Substring(0, _configuration.MaxValueSize);
+            }
+
+            return v;
+        }
+
+        protected IDictionary<string, object> FixSegmenKeysAndValues(IDictionary<string, object> segments) {
```
[the remark this answers is not in the document]
Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs]: Still  typo

File: Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs [hunk 11]
```diff
@@ -84,6 +71,21 @@ public async Task SetUserDetailsAsync(CountlyUserDetailsModel userDetailsModel)
                     throw new Exception("Accepted picture formats are .png, .gif and .jpeg");
                 }
 
+
+                userDetailsModel.Name = TrimValue(" '" + nameof(userDetailsModel.Name) + "' ", userDetailsModel.Name);
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs:75]: All these " '" + nameof(userDetailsModel.Name) + "' " are just too wasteful. Totally unnecessary 21 string concatination, not counting the ones in the call + using "nameof" 7 times.

Just use a static sting value or precompute this resulting string at module init time.

File: Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs [hunk 12]
```diff
@@ -23,6 +24,102 @@ protected AbstractBaseService(CountlyConfiguration configuration, CountlyLogHelp
...
+            }
+
+
+            return values;
+        }
+
+        protected string TrimValue(string nameOf, string v)
+        {
+            if (v.Length > _configuration.MaxValueSize) {
+                Log.Warning("[" + GetType().Name + "] TrimValue : Max allowed" + nameOf + "length is " + _configuration.MaxValueSize + ". " + v + " will be truncated.");
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs:90]: this log statement should be reworked so that it is not expected of the "nameOf" value to include whitespace at both ends. Those should already be included in the host string here.

File: Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs [hunk 13]
```diff
@@ -23,6 +24,102 @@ protected AbstractBaseService(CountlyConfiguration configuration, CountlyLogHelp
...
+                    Log.Warning("[" + GetType().Name + "] TrimKey : Max allowed value length is " + _configuration.MaxKeyLength + ". " + values[i] + " will be truncated.");
+                    values[i] = values[i].Substring(0, _configuration.MaxValueSize);
+                }
+            }
+
+
+            return values;
+        }
+
+        protected string TrimValue(string nameOf, string v)
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/AbstractBaseService.cs:87]: Rename "nameOf" to something more meaningful

File: Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs (hunk 11 above)
Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs:75]: and these passed value names are just a copy paste thing. They were not updated after pasting

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Review by ArtursKadikis — APPROVED** [review]
